internal/grid/manager.go

```diff
@@ -0,0 +1,308 @@
+		}
+
+		msg, _, err := wsutil.ReadClientData(conn)
+		if err != nil {
+			logger.LogIf(ctx, fmt.Errorf("grid: reading connect: %w", err))
+			return
+		}
+		if debugPrint {
+			fmt.Printf("%s handler: Got message, length %v\n", m.local, len(msg))
+		}
+
+		var message message
+		_, err = message.parse(msg)
+		if err != nil {
+			if debugPrint {
+				fmt.Println("parse err:", err)
+			}
+			logger.LogIf(ctx, fmt.Errorf("handleMessages: parsing connect: %w", err))
+			return
+		}
+		if message.Op != OpConnect {
+			if debugPrint {
+				fmt.Println("op err:", message.Op)
+			}
+			logger.LogIf(ctx, fmt.Errorf("handler: unexpected op: %v", message.Op))
+			return
+		}
+		var cReq connectReq
+		_, err = cReq.UnmarshalMsg(message.Payload)
+		if err != nil {
+			if debugPrint {
+				fmt.Println("handler: creq err:", err)
+			}
+			logger.LogIf(ctx, fmt.Errorf("handleMessages: parsing ConnectReq: %w", err))
+			return
+		}
+		remote := m.targets[cReq.Host]
+		if remote == nil {
+			if debugPrint {
+				fmt.Printf("%s: handler: unknown host: %v. Have %v\n", m.local, cReq.Host, m.targets)
+			}
+			logger.LogIf(ctx, fmt.Errorf("handler: unknown host: %v", cReq.Host))
+			return
+		}
+		if debugPrint {
+			fmt.Printf("handler: Got Connect Req %+v\n", cReq)
+		}
+
+		logger.LogIf(ctx, remote.handleIncoming(ctx, conn, cReq))
+	}
+}
+
+// AuthFn should provide an authentication string for the given aud.
+type AuthFn func(aud string) string
+
+// Connection will return the connection for the specified host.
+// If the host does not exist nil will be returned.
+func (m *Manager) Connection(host string) *Connection {
+	return m.targets[host]
+}
+
+// RegisterSingle will register a stateless handler that serves
+// []byte -> ([]byte, error) requests.
+// subroutes are joined with "/" to a single subroute.
+func (m *Manager) RegisterSingle(id HandlerID, h SingleHandlerFn, subroute ...string) error {
+	if !id.valid() {
+		return ErrUnknownHandler
+	}
+	s := strings.Join(subroute, "/")
+	if debugPrint {
+		fmt.Println("RegisterSingle: ", id.String(), "subroute:", s)
+	}
+
+	if len(subroute) == 0 {
+		if m.handlers.hasAny(id) && !id.isTestHandler() {
+			return ErrHandlerAlreadyExists
+		}
+
+		m.handlers.single[id] = h
+		return nil
+	}
+	subID := makeSubHandlerID(id, s)
+	if m.handlers.hasSubhandler(subID) && !id.isTestHandler() {
+		return ErrHandlerAlreadyExists
+	}
+	m.handlers.subSingle[subID] = h
+	// Copy so clients can also pick it up for other subpaths.
+	m.handlers.subSingle[makeZeroSubHandlerID(id)] = h
+	return nil
+}
+
+/*
+// RegisterStateless will register a stateless handler that serves
+// []byte -> stream of ([]byte, error) requests.
+func (m *Manager) RegisterStateless(id HandlerID, h StatelessHandler) error {
+	if !id.valid() {
+		return ErrUnknownHandler
+	}
+	if m.handlers.hasAny(id) && !id.isTestHandler() {
+		return ErrHandlerAlreadyExists
+	}
+
+	m.handlers.stateless[id] = &h
+	return nil
+}
+*/
```

Comment:
is this needed?

Reply:
@harshavardhana We want stateless, for things like traces, metrics, etc, where messages may be dropped, but we don't have to do congestion handling.

However, I decided to focus on the two other interfaces and make them fully functional/tested first.